### ISSUE TITLE
printf formatting

### ISSUE THREAD

**Issue by holder66** [stevengj/DecFP.jl#26]
Is there an easy way to print a Dec type in a decimal representation, like the %f style?

**Comment by stevengj** [comment]
I don't know, you'd have to dig through the Intel library to see if there is a more flexible function than the `__bid_to_string` function we are calling now.

**Comment by stevengj** [comment]
I looked and didn't find anything; I think we'd have to write our own function.

**Comment by jamii** [comment]
Currently:

``` julia
julia> using DecFP

julia> @sprintf("%f", Dec64(1.5))
ERROR: StackOverflowError:
 in fix_dec(::DecFP.Dec64, ::Int64) at ./printf.jl:932 (repeats 80000 times)
```

**Comment by stevengj** [comment]
The problem is that Julia base defines `fix_dec(x::Real, n::Int) = fix_dec(float(x),n)`, which gives an infinite loop if `float(x) === x`, i.e. for any floating-point type not in Base.

**Comment by ScottPJones** [comment]
Could this be fixed then by adding a definition for `Base.fix_dec(x::DecimalFloatingPoint, n::Int)`?

**Comment by stevengj** [comment]
I don't know, but I doubt that would suffice to make sprintf work.  I'm not even sure what that function is supposed to do.  You'd have to dig through the implementation.   (The whole grisu algorithm doesn't actually make a lot of sense for decimal floating point.   I suspect that you'd want to replace sprintf entirely for DecFP rather than trying to hack the existing implementation into submission.)

**Comment by stevengj** [comment]
Actually, through the base/printf.jl file in Julia, we'd want to leave the format-parsing code as-is, it looks like you just need to provide `fix_dec` and `ini_dec` (the two functions that extract the digits as an array `DIGITS`), and maybe the rest of it will work.

**Comment by ScottPJones** [comment]
One thing that really needs to be addressed before v1.0 also related to this, is that none of these functions are thread-safe (because of the use of global buffers for DIGITS in grisu.jl and _buffer in DecFP.jl, as well as global variables for things like the rounding mode and precision)

**Comment by stevengj** [comment]
(Off topic.)

**Comment by ScottPJones** [comment]
It is totally relevant to how the printf issue is addressed (as well as the issue that the DIGITS buffer in `Base.Printf` is not large enough to hold the output of all `Dec64` or a `Dec128` values) but I've added #39 to deal specifically with the rather difficult issues of making the package thread-safe.

**Comment by ValdarT** [comment]
Any update on this? Ran into this issue because it seems to break the ```describe()``` function in DataFrames.jl / DataTables.jl (if the DataFrame / DataTable has a column with DecFP.Dec64 type).

**Comment by stevengj** [comment]
@ValdarT, no one has been working on it as far as I know.

**Comment by jmkuhn** [comment]
With the addition of `trunc` from PR #43 it will be possible to do formatted printing using the [Formatting](https://github.com/JuliaIO/Formatting.jl) package.  This package uses Python style formatting instead of C style.

```
using DecFP
using Formatting

x = d"1.116"
s = fmt(".2f", x)    # => "1.12"
s = format("x is {:5.2f}", x)    # => "x is  1.12"
printfmtln("x is {:5.2f}", x)
printfmtln("x is {:.2e}", x)
```

For Julia versions 0.6 or 0.7 you will want to use the master branch of Formatting:
```
Pkg.checkout("Formatting")
```

**Comment by stevengj** [comment]
Fixed by #48.
